### PR TITLE
feat: allow dispatching/bubbling events on the `Target`

### PR DESCRIPTION
**Pull request by dgdavid**
## TL;DR

This PR allows making the `Target` aware of events in a teleported content.

## Problem

Despite the `Source` content being rendered within the `Target` _slot_, the last cannot intercept the events originated by that content. Although this is how the events work in a [React Portal](https://beta.reactjs.org/reference/react-dom/createPortal#reference), bubbling up through the React tree rather than the DOM tree, react-teleporter could have an option for _redirecting_ the event through the Target instead, as suggested in #7.

A simple use-case could be a sidebar closing itself when the user clicks on any of the actions held by it. Check out this [codesandbox project](https://codesandbox.io/p/sandbox/react-teleporter-bubbling-demo-x3wo4f) where it is illustrated that it does not work for teleported links (think in a contextual actions sections inside the sidebar).

## Proposed solution

This PR adds support for "forwarding" events to `Target` at the time of creating the teleporter. 

```js

const ContextualActions = createTeleporter({ forwardToTarget: ["onClick"] });

```

Do not get confused by the naming: the code actually dispatches a new event based on the intercepted one.

### Relaying on a `div` using the `display: contents;` CSS rule

To do so, an _intermediate_ `div` node has been added to wrap the `Source` children making possible to intercept and _forward_ events. Since this extra node can potentially break layouts because of the box model, the _`display: contents;`_ CSS rule is shipped through a (hopefully harmless) inline style. This rule "unboxes" the node, effectively _making it unobtrusive_. A sort of _native `<fragment>`_ (not invented yet) that will minimize (if not get rid of) undesired side effects.

Read more about it in the links you can find below, or in [this Codepen displaying how it works](https://codepen.io/dgdavid/pen/KKBGEyP).

<details>
<summary>Click to show/hide a list with few links to learn more about `display: contents;`</summary>

---

  * https://drafts.csswg.org/css-display/#valdef-display-contents
  * https://developer.mozilla.org/en-US/docs/Web/CSS/display#display_contents
  * https://caniuse.com/css-display-contents
  * https://bitsofco.de/how-display-contents-works/
  * https://css-tricks.com/get-ready-for-display-contents/

</details>

> **Acknowledgement:** I have been made aware of that rule thanks to that [comment](https://stackoverflow.com/a/69550164) in StackOverflow.

## Discarded solution

The only way I found to avoid the extra node used in the proposed solution is to _inject_ the event handler for dispatching the event on the `Target` to each child in the `Source` :exploding_head::exploding_head::exploding_head: 

However, I found it suboptimal not only because the handler had to be attached multiple times instead of only once, but also because of the required steps to achieve it:

* Use the [Children](https://beta.reactjs.org/reference/react/Children) API to iterate children, which could lead to fragile code according to the documentation.
* Check if the child [isValidElement](https://beta.reactjs.org/reference/react/isValidElement).
* Create props objects for the events to be forwarded, ensuring that an already attached handler for the same event is preserved. Otherwise, the full behavior will be broken.
* Clone the child element with the newly created props object by calling to [cloneElement](https://beta.reactjs.org/reference/react/cloneElement).

<details>
<summary>Click to show/hide a diff of a working code for that suboptimal solution on top of the code sent to this PR</summary>

---

```diff
 diff --git a/src/index.test.tsx b/src/index.test.tsx
index 80a4884..8a5a230 100644
--- a/src/index.test.tsx
+++ b/src/index.test.tsx
@@ -296,25 +296,6 @@ describe("teleporter", () => {
   });
 
   describe("Source", () => {
-    it("wraps its content in an 'unboxed' div", () => {
-      const Teleporter = createTeleporter();
-
-      const { getByText } = render(
-        <div>
-          <div data-testid="target">
-            <Teleporter.Target as="header" />
-          </div>
-          <div>
-            <Teleporter.Source>Hello</Teleporter.Source>
-          </div>
-        </div>
-      );
-
-      const sourceContent = getByText("Hello");
-      const sourceContentWrapper = sourceContent.closest("div");
-      expect(sourceContentWrapper).toHaveStyle("display: contents;");
-    });
-
     it("forwards given events to Target", () => {
       const Teleporter = createTeleporter({
         forwardToTarget: ["onClick", "onMouseOver"],
diff --git a/src/index.tsx b/src/index.tsx
index 2c72a45..6965148 100644
--- a/src/index.tsx
+++ b/src/index.tsx
@@ -52,17 +52,6 @@ export const createTeleporter = ({
 
   const useTargetRef = () => setElement;
 
-  const SourceWrapper: React.FC<React.DOMAttributes<HTMLElement>> = ({
-    children,
-    ...props
-  }) => {
-    return (
-      <div {...props} style={{ display: "contents" }}>
-        {children}
-      </div>
-    );
-  };
-
   const Target = ({ as: As = "div", ...props }: PropsWithAs<{}, "div">) => {
     return <As ref={setElement} {...props} />;
   };
@@ -104,16 +93,28 @@ export const createTeleporter = ({
     const handleEvent = (e: React.SyntheticEvent) =>
       element.dispatchEvent(new Event(e.type, e));
 
-    const eventHandlersProps: {
-      [key: string]: React.EventHandler<React.SyntheticEvent>;
-    } = {};
+    const newChildren = React.Children.map(children, child => {
+      if (!React.isValidElement(child)) return child;
 
-    forwardToTarget.forEach(
-      (eventName: string) => (eventHandlersProps[eventName] = handleEvent)
-    );
+      const childProps = child.props || {};
+      const newChildProps : { [key: string]: React.EventHandler<React.SyntheticEvent> } = {};
+
+      forwardToTarget.forEach((eventName: string) => {
+        if (childProps[eventName]) {
+          newChildProps[eventName] = (e) => {
+            childProps[eventName](e);
+            handleEvent(e)
+          }
+        } else {
+          newChildProps[eventName] = handleEvent;
+        }
+      });
+
+      return React.cloneElement(child, newChildProps);
+    })
 
     return ReactDOM.createPortal(
-      <SourceWrapper {...eventHandlersProps}>{children}</SourceWrapper>,
+      newChildren,
       element
     );
   };
```
</details>

## Open questions

Regarding the proposed changes, I still have some doubts

* Should that intermediate node be always there, as it is, or would it be better to only wrap Source children when there is any event to be forwarded?

* Should the code be adapted to stop the propagation of the original event once it has been "forwarded" to the Target? If so, should it be configurable or simply a consequence of forwarding the events?

## Pending tasks

- [ ] Update the README

  This is not done on purpose for:

    * postponing it to the very end, once the PR is in its final state after reviews and suggestions.

    * asking help to do so.

- [x] Update testing-library to the latest version and use `user-event`. 

     See https://testing-library.com/docs/user-event/intro#differences-from-fireevent

     ~~**Postponed for a follow-up PR.**~~ See https://github.com/gregberge/react-teleporter/pull/48

---

A big thank you to @backpackerhh for the _rubber duck debugging_ session :wink:

### CODE REVIEW

**Comment by gregberge** [comment]
I don't think the use-case is something very common, so it should not be part of the codebase but it should be possible to implement it in userland.

Also I see several caveats:

- The `div` with `display: contents` : browser support, intermediate node, breaking change
- The dispatch: dispatching React Synthetic event instead of native one, not good for every event (click position, etc...)

---

To make this possible, what do you think about supporting `children` as function?

```tsx
const forwardEvent = element => event => element.dispatch(new Event(event.type, event))

<Source>
  {element => <div onClick={forwardEvent(element)} style={{ display: "contents" }}></div>}
</Source>
```

**Comment by dgdavid** [comment]
> I don't think the use-case is something very common, so it should not be part of the codebase but it should be possible to implement it in userland.

I see. I suggested including that _bubbling the event to the target_ behavior in the library just in case could be useful for other use-cases too, due to the fact that it's optional. But I fully understand that the needed intermediate node spoils the approach.

> 
> Also I see several caveats:
> 
>     * The `div` with `display: contents` : browser support, intermediate node, breaking change

You're right! I forgot to mention that in the description along with a link to _caniuse_. Here it is https://caniuse.com/?search=contents :)

> 
>     * The dispatch: dispatching React Synthetic event instead of native one, not good for every event (click position, etc...)
> 

Just for the record, there is the `nativeEvent` attribute in the [React Synthetic Event](https://reactjs.org/docs/events.html) holding the [DOMevent](https://domevents.dev/) which might be more convenient to be dispatched in this case. I mean, going for `new Event(event.type, event.nativeEvent)`. But, honestly, I'm not sure which option is better. What's your opinion? 

---

> 
> To make this possible, what do you think about supporting `children` as function?
> 
> ```tsx
> const forwardEvent = element => event => element.dispatch(new Event(event.type, event))
> 
> <Source>
>   {element => <div onClick={forwardEvent(element)} style={{ display: "contents" }}></div>}
> </Source>
> ```

Not sure if I got it. Do you mean supporting `children` as function in react-teleporter? If so, I guess we could do something like

```tsx
import ContextualMenu from "./ContextualMenu";

function Foo() {
  // just an example that could be improved and moved to an utils module or so.
  const forwardEvents= (target, children) => )
    <div 
      style={{ display: "contents"}} 
      onClick={e => target.dispatch(e.type, e)}>
      { children }
    </div>
  );
  
  const ContextualMenuContent= () => (
    <>
      <h3>Teleported links</h3>
      <ul>
        <li>
          <a onClick={() => console.log("Link from Foo clicked!")}>
            A link from the portal
          </a>
        </li>
      </ul>
    </>
  );
  
  return (
    <>
      <div className="Foo">
        The real content for Foo around here.
      </div>
      <ContextualMenu.Source>
         { sidebar => forwardEvents(sidebar, <ContextualMenuContent />)}
      </ContextualMenu.Source>
    </>
  );
}

export default Foo;

```

**Thanks a lot** for the review and feedback!

**Comment by gregberge** [comment]
I close it since children function is now supported it can be implemented in userland.
